go/worker/common/p2p/rpc/client.go

```diff
@@ -0,0 +1,321 @@
+	) (PeerFeedback, error)
+
+	// CallMulti routes the given RPC method call to multiple peers that support the protocol based
+	// on past experience with the peers.
+	//
+	// It returns all successfully retrieved results and their corresponding PeerFeedback instances.
+	CallMulti(
+		ctx context.Context,
+		method string,
+		body, rspTyp interface{},
+		maxPeerResponseTime time.Duration,
+		maxParallelRequests uint,
+	) ([]interface{}, []PeerFeedback, error)
+}
+
+type client struct {
+	PeerManager
+
+	host       core.Host
+	protocolID protocol.ID
+	runtimeID  common.Namespace
+
+	logger *logging.Logger
+}
+
+func (c *client) Call(
+	ctx context.Context,
+	method string,
+	body, rsp interface{},
+	maxPeerResponseTime time.Duration,
+) (PeerFeedback, error) {
+	c.logger.Debug("call", "method", method)
+
+	// Prepare the request.
+	request := Request{
+		Method: method,
+		Body:   cbor.Marshal(body),
+	}
+
+	// Iterate through the prioritized list of peers and attempt to execute the request.
+	for _, peer := range c.GetBestPeers() {
+		c.logger.Debug("trying peer",
+			"method", method,
+			"peer_id", peer,
+		)
+
+		pf, err := c.call(ctx, peer, &request, rsp, maxPeerResponseTime)
+		if err != nil {
+			continue
+		}
+		return pf, nil
+	}
+
+	// No peers could be reached to service this request.
+	c.logger.Debug("no peers could be reached to service request",
+		"method", method,
+	)
+
+	return nil, fmt.Errorf("call failed on all peers")
+}
+
+func (c *client) CallMulti(
+	ctx context.Context,
+	method string,
+	body, rspTyp interface{},
+	maxPeerResponseTime time.Duration,
+	maxParallelRequests uint,
+) ([]interface{}, []PeerFeedback, error) {
+	c.logger.Debug("call multiple", "method", method)
+
+	// Prepare the request.
+	request := Request{
+		Method: method,
+		Body:   cbor.Marshal(body),
+	}
+
+	// Create a worker pool.
+	pool := workerpool.New("p2p/rpc")
+	pool.Resize(maxParallelRequests)
+	defer pool.Stop()
+
+	// Requests results from peers.
+	type result struct {
+		rsp interface{}
+		pf  PeerFeedback
+		err error
+	}
+	var resultCh []chan *result
+	for _, peer := range c.GetBestPeers() {
+		ch := make(chan *result, 1)
+		resultCh = append(resultCh, ch)
+
+		c.logger.Debug("scheduling to call peer", "peer", peer)
+
+		pool.Submit(func() {
+			c.logger.Debug("calling peer",
+				"peer", peer,
+			)
+
+			rsp := reflect.New(reflect.TypeOf(rspTyp)).Interface()
+			pf, err := c.call(ctx, peer, &request, rsp, maxPeerResponseTime)
+			ch <- &result{rsp, pf, err}
+			close(ch)
+		})
+	}
+
+	// Gather results.
+	var (
+		rsps []interface{}
+		pfs  []PeerFeedback
+	)
+	for _, ch := range resultCh {
+		select {
+		case <-ctx.Done():
+			return nil, nil, ctx.Err()
+		case result := <-ch:
+			// Ignore failed results.
+			if result.err != nil {
+				continue
+			}
+
+			rsps = append(rsps, result.rsp)
+			pfs = append(pfs, result.pf)
+		}
+	}
+	c.logger.Debug("gathered results", "num1", len(rsps), "num2", len(pfs))
```

Comment:
Maybe just `"num_results": len(rsps)`, since `rsps` and `pfs` will always be the same size?

Reply:
Oops, this is a leftover from earlier debugging, will remove and check if there's anything similar left.